README.rst
===============================
HML Monkey
===============================


.. image:: https://img.shields.io/pypi/v/hml_monkey.svg
        :target: https://pypi.python.org/pypi/hml_monkey

.. image:: https://img.shields.io/travis/mmaiers-nmdp/hml_monkey.svg
        :target: https://travis-ci.org/mmaiers-nmdp/hml_monkey

.. image:: https://readthedocs.org/projects/hml-monkey/badge/?version=latest
        :target: https://hml-monkey.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/mmaiers-nmdp/hml_monkey/shield.svg
     :target: https://pyup.io/repos/github/mmaiers-nmdp/hml_monkey/
     :alt: Updates


Lightweight HML parsing


* Free software: LGPL 3.0
* Documentation: https://hml-monkey.readthedocs.io.


Features
--------

* hml_monkey is a module that reads a directory of HML files (xml.gz) and performs imple parsing  


Warning
-------
* "Orangutan will not wear diaper" http://snltranscripts.jt.org/94/94cmonkey.phtml


HISTORY.rst
=======
History
=======

0.0.1 (2017-08-24)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

#
#    hml_monkey HML Monkey.
#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#




tests/test_hml_monkey.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
#    hml_monkey HML Monkey.
#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#



"""
test_hml_monkey
----------------------------------

Tests for `hml_monkey` module.
"""


import sys
import unittest
from contextlib import contextmanager
from click.testing import CliRunner

from hml_monkey import hml_monkey
from hml_monkey import cli



class TestHml_monkey(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_000_something(self):
        pass

    def test_command_line_interface(self):
        runner = CliRunner()
        result = runner.invoke(cli.main)
        assert result.exit_code == 0
        assert 'hml_monkey.cli.main' in result.output
        help_result = runner.invoke(cli.main, ['--help'])
        assert help_result.exit_code == 0
        assert '--help  Show this message and exit.' in help_result.output


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'mmaiers-nmdp/hml_monkey'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


hml_monkey/__init__.py
# -*- coding: utf-8 -*-

#
#    hml_monkey HML Monkey.
#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#



__author__ = """Martin Maiers"""
__email__ = 'mmaiers@nmdp.org'
__version__ = '0.0.1'


hml_monkey/gl_service.py
# -*- coding: utf-8 -*-

# validate a gl-string against
# https://gl.nmdp.org/imgt-hla/3.25.0/


import requests
import re
import string
import liftovergl

hladb_version = "3.25.0"
service_url = "http://gl.nmdp.org/imgt-hla"
noun = "genotype-list"

class GLservice:
    def __init__(self, ver=hladb_version, service_url=service_url, noun=noun):
        self.ver = ver
        self.service_url = service_url
        self.url = "{}/{}/{}".format(service_url,ver, noun)
        #self.pattern =re.compile("HLA-[A-Z]*\*[0-9:*]")
        self.pattern =re.compile("(HLA-[A-Z0-9]*)\*([0-9:]*)")
    def val(self, glstring):
        #print(self.url)
        response = requests.post(self.url, data=glstring)
        return response
    def liftover(self, glstring, source, target):
        history = liftovergl.read_history()
        gl_ids = liftovergl.mk_glids(glstring, source, history)
        return (liftovergl.mk_target(gl_ids, target, history))
    def fix(self, glstring):
        r = self.val(glstring)
        if r.status_code == 201:
            # success
            return r.text
        elif r.status_code == 400:
            # failure
            # parse text
            # e.g.: Unable to create genotype list, allele "HLA-DQB1*04:02:01:01" not a valid allele
            s = self.pattern.search(r.text)
            if s is None:
                return("No HLA")
            else: 
                loc = s.group(1)
                anum = s.group(2)
                afields = re.split(':',anum)
                numfields =  len(afields)
                oldallele = "{}*{}".format(loc, anum)
                if numfields == 3:
                     afields.append("01")
                     newanum = ':'.join(afields)
                     newallele = "{}*{}".format(loc, newanum)
                     newglstring = glstring.replace(oldallele, newallele)
                     return self.fix(newglstring)
    
                     # remove the last
                     #del afields[numfields-1]
                     #newanum = ':'.join(afields)
                     #print ("new allele: {}*{}".format(loc, newanum))
                     #newallele = "{}*{}".format(loc, newanum)
                     #newglstring = glstring.replace(oldallele, newallele)
                     #return newglstring
                elif numfields == 2:
                     # add "01:01"
                     afields.append("01")
                     newanum = ':'.join(afields)
                     newallele = "{}*{}".format(loc, newanum)
                     newglstring = glstring.replace(oldallele, newallele)
                     return self.fix(newglstring)
                else:
                     return "can't fix allele: {}*{}".format(loc, anum)
        else:
            # unknown status code
            return "unknown status: {}".format(r.status_code)
   

# run for testint
if __name__ == '__main__':  
    glv = GLservice()

    # do some liftover
    source = "3.20.0"
    target = "3.25.0"
   
    glstring = "HLA-DRB1*08:01:01/HLA-DRB1*08:01:03"
    print("{} {}".format(glstring,glv.liftover(glstring, source, target)))
    glstring = "HLA-DRB1*10:01:01+HLA-DRB1*14:04:01"
    print("{} {}".format(glstring,glv.liftover(glstring, source, target)))
    glstring = "HLA-DQB1*02:02:01:01+HLA-DQB1*04:02:01"
    print("{} {}".format(glstring,glv.fix(glstring)))


    glstring = "HLA-DQB1*02:02:01+HLA-DQB1*04:02:01"
    print("{} {}".format(glstring,glv.fix(glstring)))

    glstring = "HLA-DQB1*02:02+HLA-DQB1*04:02:01"
    print("{} {}".format(glstring,glv.fix(glstring)))

    

    


hml_monkey/hml_monkey.py
# -*- coding: utf-8 -*-

#
#    hml_monkey HML Monkey.
#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#


import sys
from lxml import etree
from io import StringIO, BytesIO
import re
import gzip
from gl_service import GLservice

glv = GLservice()

target = '3.25.0'
source = '3.20.0'
# local archive of HML
hmldir = "/Volumes/Fat/HML"
# network archive of HML
hmldir = "/Volumes/bioxover/users/mmaiers/hml/hml"

# this string seems to accompany elements
schem = "{http://schemas.nmdp.org/spec/hml/1.0.1}"

# extract the first locus from a glstring
m=re.compile("([A-Z0-9-]*)\*")
def getloc(glstring):
    l=m.search(glstring)
    if l.group(1):
        return l.group(1)
    else:
        return "NOLOC"

# dictionary of results
d={}

#for i in (range(1,3111)):
for i in (range(35,36)):
    gzfilename = "{}/{}.hml101.xml.gz".format(hmldir, i)
    fileobject = gzip.open(gzfilename, 'rb')
    print ("open {}".format(gzfilename),file=sys.stderr)
    root = etree.parse(fileobject)

    # reporting center
    reporting_center  = root.find(schem+'reporting-center')
    reporting_center_id = reporting_center.get('reporting-center-id')

    # sample
    for sample in root.findall(schem+'sample'):
        id = sample.get('id')
        # remove hyphens
        id = id.replace("-","")
        # typing
        for typing in sample.findall(schem+'typing'):
            # allele assignment
            allele_assignment = typing.find(schem+'allele-assignment')
            # glstring
            glstring = allele_assignment.find(schem+'glstring')
            gl = None
            if glstring is not None:
                if glstring.text is not None:
                   gl = glstring.text
                   loc = getloc(gl)
            else:
                loc = "NOLOC"
            #first lift it over
            lgl = glv.liftover(gl, source, target)
            lifted = 0
            if lgl != gl:
                lifted = 1

            #then fix it
            fgl = glv.fix(lgl)
            fixed = 0
            if fgl != lgl:
                fixed = 1 
            print(",".join(str(j) for j in [i, reporting_center_id, id, loc, gl, fgl,lifted, fixed]))

            if i not in d:
                d[i] = {}
            if id not in d[i]:
                d[i][id] = {}
            d[i][id][loc]= gl

for i in d.keys():
    for id in d[i].keys():
        for loc in d[i][id].keys():
            print("dump\t{}\t{}\t{}\t{}".format(i, id, loc, d[i][id][loc]))
# make a table of filename index, id, locus
# use this to construct a massive index of all HML
# run it on full archive
# make a table as the source of the 8000 cohort






hml_monkey/cli.py
# -*- coding: utf-8 -*-

#
#    hml_monkey HML Monkey.
#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
#
#    This library is free software; you can redistribute it and/or modify it
#    under the terms of the GNU Lesser General Public License as published
#    by the Free Software Foundation; either version 3 of the License, or (at
#    your option) any later version.
#
#    This library is distributed in the hope that it will be useful, but WITHOUT
#    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or
#    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#    License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with this library;  if not, write to the Free Software Foundation,
#    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.
#
#    > http://www.fsf.org/licensing/licenses/lgpl.html
#    > http://www.opensource.org/licenses/lgpl-license.php
#



import click


@click.command()
def main(args=None):
    """Console script for hml_monkey"""
    click.echo("Replace this message by putting your code into "
               "hml_monkey.cli.main")
    click.echo("See click documentation at http://click.pocoo.org/")


if __name__ == "__main__":
    main()
